### ISSUE TITLE
MumbleSetPlayerMuted / MumbleIsPlayerMuted leads to server crash

### ISSUE THREAD

**Issue by Blaxer-off** [citizenfx/fivem#2115]
Hi, 

Yesterday i was coding a script where an admin can mute a player from mumble from these natives: [https://docs.fivem.net/natives/?_0x1D5D50C2](url) and [https://docs.fivem.net/natives/?_0xCC6C2EB1](url) but when you try to mute a player it lead to a huge main stack root loop and the server crash

**Comment by AvarianKnight** [comment]
Can you provide the actual dump file?

**Comment by Blaxer-off** [comment]
There is no crash file as the server only detect the blockage and reboot by itself with the txadmin

There is what the console wrote: 

```
[              server] Error: Loop svMain seems hung! (last checkin 45 seconds ago)
[              server] Warning: svMain watchdog stack: root
> txaEvent "serverShuttingDown" "{＂delay＂:5000,＂author＂:＂txAdmin＂,＂message＂:＂Le serveur redémarre (Blocage detecté).＂}"
```

**Comment by AvarianKnight** [comment]
I can't replicate this crash, I'm on exactly the same version, try checking to see if you're doing any infinite loops anywhere because its *highly* unlikely that this native is causing this kind of hang.

**Comment by Blaxer-off** [comment]
`ESX.RegisterCommand('mute', 'owner', function(xPlayer, args, showError)
	local xTarget = ESX.GetPlayerFromId(args.playerId)
	if not xTarget then 
		TriggerClientEvent('chatMessage', xPlayer.source, '[HighLife]', {255, 0, 0}, "Joueur Hors-Ligne !")
		return 
	end

	print(MumbleIsPlayerMuted(xTarget.source))
	MumbleSetPlayerMuted(xTarget.source, not MumbleIsPlayerMuted(xTarget.source))
	Player(xTarget.source).state.isMuted = MumbleIsPlayerMuted(xTarget.source)

	local isMuted = MumbleIsPlayerMuted(xTarget.source)
	TriggerClientEvent('chatMessage', xPlayer.source, '[HighLife]', {255, 0, 0}, isMuted and "Le joueur "..xTarget.name.." est désormais mute du vocal in-game !" or "Le joueur "..xTarget.name.." n'est plus mute du vocal in-game !")
end, false, {help = "Mute/Unmute un joueur du serveur", validate = true, arguments = {
	{name = 'playerId', help = "ID du joueur", type = 'number'},
}})`

Tried again with this kind of code it crashes at the print

**Comment by AvarianKnight** [comment]
Yet again, works fine here
```lua
PMA.registerCommand('mutep', 'admin', function(ply, args, showError)
	local tgt = args.tgtPly

	local shouldBeMuted = not MumbleIsPlayerMuted(tgt.source)
	MumbleSetPlayerMuted(tgt.source, shouldBeMuted)
	Player(tgt.source).state.isMuted = shouldBeMuted

    ply.triggerEvent("chatMessage", '', { 255, 0, 0}, shouldBeMuted and ("%s was muted"):format(tgt.getPlayerName()) or ("%s was un-muted"):format(tgt.getPlayerName()))
end, false, {help = "Toggles the mute on the player", validate = true, arguments = {
	{name = 'tgtPly', help = "The target playerId", type = 'player'},
}})
```
![image](https://github.com/citizenfx/fivem/assets/13681939/ef83d330-8e09-4036-beec-17f02d83b17a)

**Comment by gottfriedleibniz** [comment]
In Blaxer-off's case it is likely that `xTarget.source` is a string and those Mumble native calls will iterate over the entire client list. Whether that could lead to a data race with `Client_add` (or something) is another question.

**Comment by AvarianKnight** [comment]
You're right, these are also the only natives to try and use playerSrc as an int (which is my fault ☹️).

I'm not entirely sure how to fix this without breaking backwards compatibility with C# though, in this case I'm not sure it should matter

**Comment by thelindat** [comment]
> it is likely that `xTarget.source` is a string

ESX stores `player.source` as a number.

**Comment by AvarianKnight** [comment]
> > it is likely that `xTarget.source` is a string
> 
> ESX stores `player.source` as a number.

Likely modified version of ESX?

**Comment by gottfriedleibniz** [comment]
> You're right, sending a string source will cause the server to hang

Have you actually caused a server to hang? And if so would you happen to have a stack-trace/dump? 

The string theory was a roundabout way to theorize that invalid inputs seemed especially data-racey as clients may also be added/removed+free'd from the client list (as `iterate_safe` caches `next`). It's not something I've investigated though.

**Comment by AvarianKnight** [comment]
No, sadly, but the int is properly overflowing, so I'm leaning towards to being a data race.

![image](https://github.com/citizenfx/fivem/assets/13681939/22fd698f-6e43-49fb-bfbc-a632b52dc96e)
![image](https://github.com/citizenfx/fivem/assets/13681939/46aa0faf-ae44-4d0b-ad8a-f5fd1d877608)

**Comment by Blaxer-off** [comment]
Hi, is it possble to get a build with your modified file so i can test it out ?

**Comment by SeaLife** [comment]
Hey! Our server is also hanging on the mute command.

```lua
RegisterNetEvent('pma-voice:toggle-mute', function()
    local source = source
    local isMuted = Player(source).state.muted
    local selfMuted = Player(source).state.selfMuted

    if isMuted then print('Cant mute ' .. source .. ' because the player is muted') return end -- dont change state if state var is set, will be set by playerState

    if not MumbleIsPlayerMuted(source) then
        MumbleSetPlayerMuted(source, true)
        Player(source).state.selfMuted = true
        print('Muted Player ' .. source .. ' due to local mute')
        TriggerClientEvent('SaltyChat_PluginStateChanged', source, 0)
    else
        MumbleSetPlayerMuted(source, false)
        Player(source).state.selfMuted = false
        print('Unmuted player ' .. source .. ' due to local mute')
        TriggerClientEvent('SaltyChat_PluginStateChanged', source, 2)
    end
end)
```

This event is called in a onjustpressed event on the client with a cooldown of 1.5seconds, so its not a loop calling it.

We'r on the latest fxserver as of now on Linux using an external Mumble server (dockerized, with latest tag), we tried with internal mumble without luck. This issue is not 100% persistent

Server log:
```
[              server] Error: Loop svMain seems hung! (last checkin 45 seconds ago)
[              server] Warning: svMain watchdog stack: root
```

**Comment by AvarianKnight** [comment]
How exactly did you test this with an external server?

The built in server doesn't have any ways of interacting with an external server so your mute calls would be called on the local server.

Unless you can get a crash dump on a windows machine I'm not sure how much can be helped here either :(

**Comment by SeaLife** [comment]
> How exactly did you test this with an external server?
> 
> The built in server doesn't have any ways of interacting with an external server so your mute calls would be called on the local server.
> 
> Unless you can get a crash dump on a windows machine I'm not sure how much can be helped here either :(

So we have this issue on 15015 and 16501 but 16237 is running fine. The Server would also crush on the built-in /muteply command (so no custom code required on our end). We dont have the abbility to test this on windows on short-term tho, so.. We stick to 16237 for now.

//EDIT: we had a crash at 11AM today (MESZ) on 16237

```
[script:avrp_evidence] Warning: Comitted 2 logs to loki
[              server] Error: Loop svMain seems hung! (last checkin 45 seconds ago)
[              server] Warning: svMain watchdog stack: pma-voice-additions: event pma-voice:toggle-mute <- root
║             TXADMIN║ Restarting server: Server is not responding (HB:1m|HC:32s).
║             TXADMIN║ - Stopped receiving HeartBeats 1m ago.
║             TXADMIN║ - HealthChecks failing for the past 32s.
║             TXADMIN║ - HealthCheck Request error: Timeout awaiting 'request' for 1500ms
║             TXADMIN║ txaEvent "serverShuttingDown" "{\"delay\":5000,\"author\":\"txAdmin\",\"message\":\"Server Neustart (der server muss neu gestartet werden, bitte verbinde dich erneut).\"}"
```

The full Code is:

Client
```lua
local function ToggleLocalMicrophone()
    if LocalPlayer.state.muted then return end
    TriggerServerEvent('pma-voice:toggle-mute')
end

lib.addKeybind({
    defaultKey = 'Y',
    description = 'Mutet dein Mikrofon',
    name = 'mutemic',
    onPressed = function (self)
        ToggleLocalMicrophone()
    end
})
```

Server
```lua
RegisterNetEvent('pma-voice:toggle-mute', function()
    local source = tonumber(source) -- ensure source is a number
    if not source then return end

    local isMuted = Player(source).state.muted
    local selfMuted = Player(source).state.selfMuted

    if isMuted then print('Cant mute ' .. source .. ' because the player is muted') return end -- dont change state if state var is set, will be set by playerState

    if not MumbleIsPlayerMuted(source) then
        MumbleSetPlayerMuted(source, true)
        Player(source).state.selfMuted = true
        print('Muted Player ' .. source .. ' due to local mute')
        TriggerClientEvent('SaltyChat_PluginStateChanged', source, 0)
    else
        MumbleSetPlayerMuted(source, false)
        Player(source).state.selfMuted = false
        print('Unmuted player ' .. source .. ' due to local mute')
        TriggerClientEvent('SaltyChat_PluginStateChanged', source, 2)
    end
end)
```

**Comment by AvarianKnight** [comment]
My guess is this is crashing in `MUMBLE_IS_PLAYER_MUTED`, the function it calls is missing a return value which (probably?) is hitting undefined behavior.

This is just a shot in the dark because everything else here looks fine.

**Comment by AvarianKnight** [comment]
Using `MumbleIsPlayerMuted` with a player that isn't connected to the mumble server does indeed cause this crash on Linux, it spikes the CPU usage to max and hangs. On Windows it looks to return invalid data too, but doesn't crash.

**Comment by AvarianKnight** [comment]
@SeaLife Can you try the latest version and see if this still crashes?
